### PR TITLE
Feature/cpp aql reverse

### PR DESCRIPTION
**Pull request by dothebart**

### CODE REVIEW

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/157/

**Comment by dothebart** [comment]
Errors in the testruns are unrelated to these changes.

File: arangod/Aql/Functions.cpp [hunk 1]
```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
...
+  }
+
+  transaction::StringBufferLeaser buffer(trx);
+  arangodb::basics::VPackStringBufferAdapter adapter(buffer->stringBuffer());
+
+  AppendAsString(trx, adapter, value);
+  UnicodeString unicodeStr(buffer->c_str(), static_cast<int32_t>(buffer->length()));
+
+  auto locale = LanguageFeature::instance()->getLocale();
+  std::vector<std::unique_ptr<StringSearch>> searchVec;
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: maybe add a note here that the iterator can not be copied

File: arangod/Aql/Functions.cpp [hunk 2]
```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
...
+    searchVec.push_back(std::make_unique<StringSearch>(searchStr, unicodeStr, locale, nullptr, status));
+  }
+  
+  std::vector<std::pair<int32_t, int32_t>> srchResultPtrs;
+  std::string utf8;
+  srchResultPtrs.reserve(match.size());
+  for (auto& search : searchVec) {
+    // We now find the first hit for each search string. 
+    auto pos = search->first(status);
+    int32_t len;
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: Should this be initialized for the case that  `POS == USEARCH_DONE`
Will pair that have USEARCH_DONE in `.first` be ignored later on? If so why add them in the first place?

File: arangod/Aql/Functions.cpp [hunk 3]
```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
...
+    int i = 0;
+    for (auto resultPair : srchResultPtrs) {
+      // We locate the nearest matching search result.
+      int32_t thisPos;
+      thisPos = resultPair.first;
+      if ((pos == USEARCH_DONE) || (pos > thisPos)) {
+        if (thisPos != USEARCH_DONE) {
+          pos = thisPos;
+          which = i;
+          mLen = resultPair.second;
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: looks like this could become an arbitrary value

File: arangod/Aql/Functions.cpp [hunk 4]
```diff
@@ -1288,6 +1347,232 @@ AqlValue Functions::Substring(arangodb::aql::Query* query,
...
+    if (allFound) {
+      break;
+    }
+  }
+  // Append from the last found:
+  result.append(unicodeStr, lastStart, unicodeStr.length() - lastStart);
+  
+  result.toUTF8String(utf8);
+  return AqlValue(utf8);
+}
```
Comment (ObiWahn) [inline arangod/Aql/Functions.cpp]: I think the function is too long and should be split (No need to change it this time)

File: arangod/Aql/Functions.cpp [hunk 5]
```diff
@@ -925,6 +926,64 @@ AqlValue Functions::FindLast(arangodb::aql::Query* query,
...
+    array.reserve(slice.length());
+    for (auto const& it : VPackArrayIterator(slice)) {
+      array.push_back(it);
+    }
+    std::reverse(std::begin(array), std::end(array));
+    
+    builder->openArray();
+    for (auto const &it : array) {
+      if (it.isCustom()) {
+        builder->add(VPackValue(trx->extractIdString(slice)));/// TODO
```
[the remark this answers is not in the document]
Comment (ObiWahn) [reply arangod/Aql/Functions.cpp]: Extract from slice or it?

[the remark this answers is not in the document]
Comment (dothebart) [reply arangod/Aql/Functions.cpp]: ok, I don't think we want to convert anything, so this should prably just go away.

File: arangod/Aql/Functions.cpp (hunk 2 above)
Comment (dothebart) [reply arangod/Aql/Functions.cpp]: the number in the vectors has to be identical, so counters can be easily be used in the loop below.

**Comment by dothebart** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/170/

File: js/common/bootstrap/errors.js [hunk 6]
```diff
@@ -38,6 +38,7 @@
     "ERROR_SHUTTING_DOWN"          : { "code" : 30, "message" : "shutdown in progress" },
     "ERROR_ONLY_ENTERPRISE"        : { "code" : 31, "message" : "only enterprise version" },
     "ERROR_RESOURCE_LIMIT"         : { "code" : 32, "message" : "resource limit exceeded" },
+    "ERROR_ARANGO_ICU_ERROR"       : { "code" : 33, "message" : "icu error: %s" },
```
Comment (ObiWahn) [inline js/common/bootstrap/errors.js]: remove the _ERROR suffix as it is already an error

**Pull request merged**